pkg/util/startup/retry.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package startup
+
+import (
+	"context"
+	"reflect"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+type startupRetryKey struct{}
+
+var startupRetryOpts = retry.Options{
+	InitialBackoff: 100 * time.Millisecond,
+	MaxBackoff:     3 * time.Second,
+	Multiplier:     2,
+}
+
+// runningStartup is a counter showing number of concurrently running server
+// startups. if it is zero, then startups are finished and we should not try
+// to run startup retry checks.
+var runningStartup atomic.Int32
+
+// RetryContext adds startup tag to allow kv access from startup path.
+func RetryContext(ctx context.Context) context.Context {
```

Comment:
Name suggestion: `WithoutChecks`

to make clear that you get a context that avoids all of the checks in this package. I.e. you opt into knowing better than the assertions.
A clearer comment to that effect would also be helpful.